### PR TITLE
Tests now fail when an error occurs.

### PR DESCRIPTION
**Pull request by ezzatron**
Continuing from #175, this PR further improves the error handling in Peridot.

The biggest change is that when an error occurs in a test, that test will now fail. This new behavior will honor the `error_reporting` setting, meaning that systems that typically ignore errors of a given severity can still be tested under Peridot. This is the same underlying mechanism used by the error control operator (`@`), so that will also work as intended.

Regardless of the `error_reporting` level, a Peridot `error` event will still be emitted. It will be up to individual listener implementations to decide whether to honor the `error_reporting` level.

Another change is to the behavior of tests in which multiple failures occur. Previously, if a test failed AND its teardown also failed, the teardown failure would "overwrite" the test failure. This behavior has changed so that the first failure is always the failure that is reported. This has particular significance given that a test can now fail because of an error, and then continue on to fail because of an assertion failure or other exception.

I also added support for PHP 7 [engine exceptions](http://php.net/language.errors.php7). Note that I have polyfilled the existence of [Throwable](http://php.net/throwable) and [Error](http://php.net/error), so that the tests may also be run under PHP 5.

This PR *should* also solve #158 and #132.

### CODE REVIEW

**Comment by ezzatron** [comment]
Hmm, failing because `set_error_handler(null)` doesn't work until PHP 5.5. Hang tight.

**Comment by ezzatron** [comment]
Please don't merge, something's still not right.

**Comment by ezzatron** [comment]
@jmalloc FYI

File: composer.json [hunk 1]
```diff
@@ -30,5 +30,10 @@
       "Peridot\\": "src/"
     }
   },
+  "autoload-dev": {
```
Comment (brianium) [inline composer.json]: Would this be ignored by anyone requiring Peridot via composer? My concern here is that our build will pass as it runs `bin/peridot` but anyone requiring via composer would not have this dependency loaded. @ezzatron can you confirm?

File: src/Core/Test.php [hunk 2]
```diff
@@ -52,13 +54,21 @@ public function run(TestResult $result)
     protected function executeTest(TestResult $result)
     {
         $action = ['passTest', $this];
+        $handlers = $this->handleErrors($result, $action);
         try {
             $this->runSetup();
             call_user_func_array($this->getDefinition(), $this->getDefinitionArguments());
+        } catch (Throwable $e) {
+            if ('passTest' === $action[0]) {
```
Comment (brianium) [inline src/Core/Test.php]: This pattern seems to be repeated in a few places, any way we can pull this out into a common method/function? Does it seem worth it?

Something like `failIfPassing(array $action): array`?

File: src/Core/Test.php [hunk 3]
```diff
@@ -88,11 +98,88 @@ protected function runTearDown(TestResult $result, $action)
...
+     * @param array      &$action
+     *
+     * @return array
+     */
+    protected function handleErrors(TestResult $result, &$action)
+    {
+        // first, unwind the current handler stack for later restoration
+        $handlers = [];
+
+        do {
```
Comment (brianium) [inline src/Core/Test.php]: Wondering if we could extract this pattern of walking error handlers as well:

```php
function walkErrorHandlers(callable $cb) {
  do {
    if ($handler = set_error_handler(function() {})) {
      $cb($handler);
      restore_error_handler();
    }
    
    restore_error_handler();
  } while ($handler);
}
```

Thoughts?

**Comment by brianium** [comment]
@ezzatron left a few comments, but overall I dig the approach. Seems like it will be an overall improvement. Good work!

My one question is a performance related one, and as such it might be paranoid/too early to tell. **Do you think we have anything to worry about with unwinding and affecting error handling on every test run?** Did you see any noticeable slow downs while running tests during development? Comparing some builds on Travis, I saw that the PR tests ran a bit slower than previous builds. This could be any number of things so I am not putting too much stock in it. Maybe profiling a run with the new error handling stuff would show us if there are any bottlenecks we need to be concerned with.

Thoughts?

**Comment by ezzatron** [comment]
@brianium Performance is a genuine concern here, I agree. I didn't notice a huge difference running the Peridot suite, but I am testing under PHP 7 primarily, and everything is super fast there anyway.

I can probably avoid unwinding the stack each time, and only deal with the "previous" handler. It would have to be faster than unwinding and re-adding each time. The downside is that if the system under test adds error handlers, they won't get cleaned up by Peridot. I don't think that's really a problem, it's probably an edge-case anyway. I think I'll update the PR to take this approach.

As a general note, reducing the number of function calls will usually help with performance, especially under PHP 5 with xdebug / code coverage. Having made PRs to PHPUnit before, they're always very cautious about introducing extra calls. If you want to see some significant performance improvements, the first thing I would do is to use less calls in `Test::run()`. You have to draw the line somewhere though; collapsing methods into each other to avoid additional calls quickly gets ugly ;)

**Comment by ezzatron** [comment]
Okay, I've finished updating things for now.

It's pretty hard to know if there is any real performance issue. I looked at the Travis run times, but they seem to change like the wind, and I don't think they're a good indicator. A quick cachegrind analysis didn't turn up any obvious problems. Running it manually, I can't perceive any significant difference in performance.

Perhaps the most telling result I've seen comes from running HTML code coverage on the peridot suite itself under PHP 5.6. Here's a quick result set:

master (seconds) | fail-on-error (seconds)
-----------------|------------------------
11.58            | 11.86
11.46            | 11.90
11.77            | 11.90
11.54            | 11.72


master average   | fail-on-error average
-----------------|------------------------
11.59            | 11.85

Basically indicating that yes, it is *slightly* slower. But that's sort of expected, given that it's actually doing more now.

File: composer.json (hunk 1 above)
Comment (ezzatron) [inline composer.json]: It will be ignored, yes. That's by design, as the polyfill is only needed for test coverage under PHP 5. I have tested this branch in another project under both PHP 5 and 7, and it works fine.

File: src/Core/Test.php (hunk 2 above)
Comment (ezzatron) [inline src/Core/Test.php]: Done.

File: src/Core/Test.php [hunk 4]
```diff
@@ -81,18 +88,79 @@ protected function runSetup()
...
      */
-    protected function runTearDown(TestResult $result, $action)
+    protected function runTearDown(TestResult $result, array $action)
     {
         $this->forEachNodeBottomUp(function (TestInterface $test) use ($result, &$action) {
             $tearDowns = $test->getTearDownFunctions();
             foreach ($tearDowns as $tearDown) {
                 try {
                     $tearDown();
+                } catch (Throwable $e) {
```
Comment (brianium) [inline src/Core/Test.php]: I think this is where I got hung up on that `autoload-dev` business. Type hinting a class that will never exist in a PHP 5 environment (when installed via composer). Since it's part of a `try-catch` block, it's just ignored in those environments.

**Comment by brianium** [comment]
@ezzatron awesome! I think we are just about there. We took an ever so slight [dip](https://scrutinizer-ci.com/g/peridot-php/peridot/inspections/0eb9baff-e415-444c-b921-0fa44cc414c0/issues/files/src/Core/Test.php?status=new&orderField=path&order=asc&honorSelectedPaths=0) on the Scrutinizer build because of the polyfill.

Wondering if its as easy as adding that fixture file to the paths that are linted via `.scrutinizer.yml`. I realize a polyfill isn't a typical thing, and the dip isn't that large - so if you spend more than 5 minutes and can't get Scrutinizer to play nicely then I would be willing to punt on it and just merge this.

Super excited about this. Thanks again!

**Comment by ezzatron** [comment]
I managed to at least get rid of the Scrutinizer "issues". There's a `dependency_paths` option that it treats as if it were external library code, which is perfect for this case. Still went down a little in quality because of the complexity of `handleErrors()`, I think.

**Comment by brianium** [comment]
@ezzatron 
post-48776-key-and-peele-noice-gif-imgur-fvbr

**Comment by brianium** [comment]
ill tag a new minor version with this stuff in shortly

**Comment by ezzatron** [comment]
Haha! Noice :sunglasses:
